File: svc/ctrl/worker/deploy/parse_fork_ref_test.go
package deploy

import (
	"testing"

	"github.com/stretchr/testify/require"
)

func TestParseForkRef(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name       string
		branch     string
		wantOwner  string
		wantBranch string
		wantOK     bool
	}{
		{
			name:       "fork ref with slash in branch",
			branch:     "alice-dev:feat/redesign-navbar",
			wantOwner:  "alice-dev",
			wantBranch: "feat/redesign-navbar",
			wantOK:     true,
		},
		{
			name:       "fork ref with simple branch",
			branch:     "contributor:feature-branch",
			wantOwner:  "contributor",
			wantBranch: "feature-branch",
			wantOK:     true,
		},
		{
			name:   "plain branch",
			branch: "main",
			wantOK: false,
		},
		{
			name:   "branch with slash",
			branch: "feature/something",
			wantOK: false,
		},
		{
			name:   "empty string",
			branch: "",
			wantOK: false,
		},
		{
			name:   "empty owner",
			branch: ":branch",
			wantOK: false,
		},
		{
			name:   "empty branch after colon",
			branch: "owner:",
			wantOK: false,
		},
		{
			name:   "slash before colon rejected",
			branch: "org/owner:branch",
			wantOK: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()
			owner, branch, ok := parseForkRef(tt.branch)
			require.Equal(t, tt.wantOK, ok)
			if ok {
				require.Equal(t, tt.wantOwner, owner)
				require.Equal(t, tt.wantBranch, branch)
			}
		})
	}
}

File: web/apps/dashboard/app/(app)/[workspaceSlug]/projects/[projectId]/(overview)/navigations/parse-fork-ref.test.ts
import { describe, expect, it } from "vitest";
import { parseForkRef } from "./parse-fork-ref";

describe("parseForkRef", () => {
  it("parses a fork reference", () => {
    expect(parseForkRef("contributor:feature-branch")).toEqual({
      forkOwner: "contributor",
      branch: "feature-branch",
    });
  });

  it("parses a fork reference with slash in branch", () => {
    expect(parseForkRef("alice-dev:feat/redesign-navbar")).toEqual({
      forkOwner: "alice-dev",
      branch: "feat/redesign-navbar",
    });
  });

  it("rejects an https URL", () => {
    expect(parseForkRef("https://github.com/owner/repo")).toBeNull();
  });

  it("rejects an http URL with colon-port", () => {
    expect(parseForkRef("http://example.com:8080")).toBeNull();
  });

  it("rejects a plain branch name", () => {
    expect(parseForkRef("main")).toBeNull();
  });

  it("rejects a branch with slash but no colon", () => {
    expect(parseForkRef("feature/something")).toBeNull();
  });

  it("rejects an empty string", () => {
    expect(parseForkRef("")).toBeNull();
  });

  it("rejects an empty owner", () => {
    expect(parseForkRef(":branch")).toBeNull();
  });

  it("rejects an empty branch", () => {
    expect(parseForkRef("owner:")).toBeNull();
  });

  it("rejects a slash before colon", () => {
    expect(parseForkRef("org/owner:branch")).toBeNull();
  });
});
